### PR TITLE
New HLR From pt 2

### PR DESCRIPTION
**Pull request by michelpmcdonald**
Part 2 of new hlr form, had to split into multiple prs, see this pr:
https://github.com/department-of-veterans-affairs/vets-api/pull/17353

## Summary

Added new revision of the official VA HLR pdf form.
We are calling the new HLR form v4.
Added the new v4 HLR form to version 2 of our higher level reviews api(API version is not tied to only one specific VA form version. Each version of or API supports multiple versions of the the PDF HLR form)

*This work is behind a feature toggle: decision_review_hlr_form_v4_enabled

No new API revision\release is needed.  Generating the filled in HLR form is consider an internal implementation detail and we make no representation to consumers which specific version of the HLR form used

## Related issue(s)

[API-36087](https://jira.devops.va.gov/browse/API-36087)

## Testing done

Unit testing is limited when generating PDFs.  The unit test don't work when comparing a API generated PDF with an expected PDF example on k8's.

Manually generated PDFs and checked to make sure fields as are expected.

*** Actual unit test to follow in separate PR, avoiding repo PR line count limit

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
Decision Reviews HLR v2, when enabled we will generate the latest version of the form to send to central mail.
Not consider an API breaking change, the form we generate is mostly an internal artifact, we make no statements around which version of the HLR PDF we use to consumers.

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution

### CODE REVIEW

File: modules/appeals_api/spec/services/appeals_api/pdf_construction/generator/generator_hlr_spec.rb
Comment (rmtolmach) [inline modules/appeals_api/spec/services/appeals_api/pdf_construction/generator/generator_hlr_spec.rb:157]: The `k8s` branch is no longer used for vets-api (we deploy from `master` now) so this comment can go 😄 

Also, the test right above this one ☝️, can it be uncommented out now?

File: modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/form_fields.rb
Comment (rmtolmach) [inline modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/form_fields.rb:9]: I'm not quite sure how to get around disabling these cops. But one option would be creating constants for some of the repeated parts. Like this:

```ruby
  FORM_PATH = 'form1[0].#subform['.freeze
  VETERAN_PATH = "#{FORM_PATH}2].Veterans_".freeze
  CLAIMANT_PATH = "#{FORM_PATH}2].Claimants_".freeze
  ADDRESS_PATH = "#{FORM_PATH}2].CurrentMailingAddress_".freeze
  PHONE_PATH = "#{FORM_PATH}2].Telephone_".freeze
  DATE_PATH = "#{FORM_PATH}4].Date_Signed_".freeze

  FIELD_NAMES = {
    veteran_middle_initial: "#{VETERAN_PATH}Middle_Initial1[0]",
    veteran_ssn_first_three: "#{VETERAN_PATH}SocialSecurityNumber_FirstThreeNumbers[0]",
   ...
    veteran_dob_day: "#{FORM_PATH}2].DOBday[0]",
    veteran_dob_month: "#{FORM_PATH}2].DOBmonth[0]",
   ...
    veteran_state_code: "#{ADDRESS_PATH}StateOrProvince[0]",
   ...
```
That seems a bit cleaner to me. What do you think?

File: modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/structure.rb
Comment (rmtolmach) [inline modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/structure.rb:55]: Is this method necessary?

File: modules/appeals_api/spec/services/appeals_api/pdf_construction/generator/generator_hlr_spec.rb
Comment (michelpmcdonald) [reply modules/appeals_api/spec/services/appeals_api/pdf_construction/generator/generator_hlr_spec.rb:157]: The plan is to re-evaluate these tests after this PR get merged....in the past these test have been fragile, sometimes passing, sometimes not, and we don't want to conflate the new version of the form changes with these failing test.  We do run them locally when we change the from generation code.

(We think the issue with these test is slightly different behavior of the PDFtk library on the k8 linux vs our OS X development machines)

File: modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/form_fields.rb
Comment (michelpmcdonald) [reply modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/form_fields.rb:9]: IMHO, that would make easily reading the complete name PDF field names(not our names, VA HLR form) a little bit tricky given the multiple levels of indirection......the use case for reading this code most likely would be for a developer attempting to find that field name string within the PDF(i use firefox to look/find the PDF field names within the PDF)

File: modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/structure.rb
Comment (michelpmcdonald) [reply modules/appeals_api/app/services/appeals_api/pdf_construction/higher_level_review/v4/structure.rb:55]: Yes, each specific form like this (new VA HLR v4 form) is called by by our generic AppealsApi::PdfConstruction::Generator calls this for all of our specific forms implementations like this one.

**Comment by mathisto** [comment]
I've taken a look at what it would take to refactor the existing code to meet the new guidelines, and it's going to be a pretty time-consuming process that's beyond the scope of this particular PR.

To make sure we can address these changes properly, I've gone ahead and created a separate ticket, [API-39184](https://jira.devops.va.gov/browse/API-39184), to track the refactoring work. This way, we can plan it out and allocate the necessary resources without holding up this current PR.

@rmtolmach, Given the timeline of events and the amount of effort needed to get everything in line with the new reek requirements, I'm hoping we can consider a dispensation for this specific PR. If you have any questions or concerns, please rope in either me or @cilestin.

**Review by cilestin — APPROVED** [review]

**Comment by rmtolmach** [comment]
Hey @mathisto I'm not on support anymore so I've added backend-review-group. @ericboehs will see it! :)

**Comment by ericboehs** [comment]
> Hey @mathisto I'm not on support anymore so I've added backend-review-group. @ericboehs will see it! :)

Unfortunately this has "omit-backend-review" on it so I missed it. Thankfully I was scrolling through github notifications on my phone and saw this! Taking a look.

**Review by ericboehs — APPROVED** [review]

**Pull request merged**
